kratos/processes/apply_ray_casting_process.cpp

```diff
@@ -0,0 +1,485 @@
+                // considering the very near points as the same points
+                if (std::abs(i_begin->first - i_intersection->first) > epsilon) // if the hit points are far enough they are not the same
+                    *(++i_intersection) = *i_begin;
+            }
+            rIntersections.resize((++i_intersection) - rIntersections.begin());
+        }
+	}
+
+	template<std::size_t TDim>
+	int  ApplyRayCastingProcess<TDim>::GetCellIntersections(
+		OctreeType::cell_type* cell,
+		const double* ray,
+		OctreeType::key_type* ray_key,
+		const unsigned int direction,
+		std::vector<std::pair<double, Element::GeometryType*> > &rIntersections)
+	{
+		//This function passes the ray through the cell and gives the hit point to all objects in its way
+		//ray is of dimension (3) normalized in (0,1)^3 space
+		// direction can be 0,1,2 which are x,y and z respectively
+		typedef OctreeType::cell_type::object_container_type object_container_type;
+		object_container_type* objects = (cell->pGetObjects());
+
+		// There are no intersection in empty cells
+		if (objects->empty()){
+			return 0;
+		}
+
+		// Calculating the two extreme of the ray segment inside the cell
+		double ray_point1[3] = {ray[0], ray[1], ray[2]};
+		double ray_point2[3] = {ray[0], ray[1], ray[2]};
+		double normalized_coordinate;
+
+		OctreeType* pOctree = mpFindIntersectedObjectsProcess->GetOctreePointer();
+		pOctree->CalculateCoordinateNormalized(ray_key[direction], normalized_coordinate);
+		ray_point1[direction] = normalized_coordinate;
+		ray_point2[direction] = ray_point1[direction] + pOctree->CalcSizeNormalized(cell);
+		pOctree->ScaleBackToOriginalCoordinate(ray_point1);
+		pOctree->ScaleBackToOriginalCoordinate(ray_point2);
+
+		// This is a workaround to avoid the z-component in 2D
+		if (TDim == 2){
+			ray_point1[2] = 0.0;
+			ray_point2[2] = 0.0;
+		}
+
+		for (object_container_type::iterator i_object = objects->begin(); i_object != objects->end(); ++i_object){
+			double intersection[3]={0.0, 0.0, 0.0};
+			const int is_intersected = ComputeRayIntersection((*i_object)->GetGeometry(), ray_point1, ray_point2, intersection);
+			if (is_intersected == 1){ // There is an intersection but not coplanar
+				rIntersections.push_back(std::pair<double, Element::GeometryType*>(intersection[direction], &((*i_object)->GetGeometry())));
+			}
+		}
+
+		return 0;
+	}
+
+	template<>
+	int ApplyRayCastingProcess<2>::ComputeRayIntersection(
+		Element::GeometryType& rGeometry,
+        const double* pRayPoint1,
+        const double* pRayPoint2,
+        double* pIntersectionPoint)
+	{
+		// Auxiliar arrays 
+		array_1d<double,3> ray_pt_1;
+		array_1d<double,3> ray_pt_2;
+		for (unsigned int i = 0; i < 3; ++i){
+			ray_pt_1[i] = pRayPoint1[i];
+			ray_pt_2[i] = pRayPoint2[i];
+		}
+
+		// Call the line - line intersection util
+		array_1d<double,3> int_pt = ZeroVector(3);
+		const double tolerance = 1.0e-6*rGeometry.Length();
```

Comment:
This is just to note that this tolerance should be changed. This has been tried in commit 98c1b250efc1b5ce713b00a23ee5d0fe6904ef04 .

Reply:
But I concluded that we change the tolerance in separate PR to make the refactoring without behavior change

Reply:
Yes, we agreed on that. This was just a reminder :wink: